File: web/lib/scrub.ts
// Email and IP addresses do not reach this container's log, and neither do the
// bytes that let one line pretend to be two.
//
// A PORT, NOT A SECOND OPINION. The original is api/internal/logx/scrub.go and
// it is the one that has been fuzzed. Everything structural here — one
// left-to-right pass, longest candidate first and then shorter, the length cap,
// a marker built only from characters a domain may contain — is copied because
// each of those is a bug the fuzzer found in Go, and a rewrite from the
// description would find them again the hard way. The three inputs that
// disproved the Go version have tests here under the same names.
//
// WHY WEB NEEDS THIS AT ALL, given that web logs no form contents
//
// Because ADR 0037 does not divide the work by "our fields" and "their fields",
// it divides it by WHOSE WORDS THEY ARE. The scrubber is for foreign text — and
// web gets some the moment it fetches:
//
//     TypeError: fetch failed
//       cause: Error: connect ECONNREFUSED 172.18.0.3:8080
//
// That is undici's wording with an address in it, in the error line of
// serverFetch, and ADR 0035 says the api is briefly gone during step 3 of every
// rollout. Without a filter, every rollout writes container addresses into the
// log. Discipline at the call site cannot see it: the address is not a field
// anybody here chose.
//
// WHAT IS DELIBERATELY NOT PORTED
//
// The `selfAuthored` exemption. Go has exactly one entry, `message_id`, because
// an RFC 5322 Message-ID is genuinely address-shaped and internal/contact mints
// it. Web writes no such field, and a list guarding a case that does not exist
// is what "Maß halten" warns about. If web ever logs one, that is when the list
// appears — and it is a claim about the value, not a convenience for the key.

// What a redacted value is replaced with. A marker rather than an empty string:
// "this line had an address and it was removed" is a different fact from "this
// line had no address", and only the first tells a reader the filter runs.
//
// Built only from characters that may appear inside a domain, and always
// containing a hyphen — not cosmetic, and the Go fuzzer is why it is written
// down. With brackets around it, redacting an address in the middle of a longer
// token ENDED the domain scan at the bracket and left the text before it looking
// like a valid domain, so a second pass redacted more than the first. A marker
// the scanner reads straight through cannot move a boundary, and the hyphen
// keeps the result from ever being a well-formed suffix.
const REDACTED_EMAIL = "redacted-email";
const REDACTED_IP = "redacted-ip";

// Bounds one candidate. See matchAddr for what it costs and why.
const MAX_ADDR_LEN = 64;

// C0 and DEL. Multi-byte characters are left alone: every continuation unit is
// above this range, so this cannot cut a character in half.
const CONTROL = /[\u0000-\u001f\u007f]/g;

/**
 * Removes control characters, email addresses and IP addresses from one string.
 */
export function scrub(s: string): string {
  return redactAddresses(stripControl(s));
}

/**
 * Removes the characters that let one log line pretend to be two.
 *
 * `JSON.stringify` already escapes them, and that is measured rather than
 * assumed on the Go side: a newline in a path comes out as an escape INSIDE the
 * string and the line stays one line. So this is not what makes forging
 * impossible today — it is what keeps it impossible when the writer changes. F2
 * sends these lines through Alloy into Loki and F11 adds a second producer, and
 * "we are safe because of how the encoder at the bottom happens to behave" is a
 * guarantee living in a different file from the values it protects.
 *
 * A space, not a deletion: two tokens run together read as one value that was
 * never there.
 */
export function stripControl(s: string): string {
  return s.replace(CONTROL, " ");
}

/**
 * Renders any thrown value as the text worth logging.
 *
 * The cause chain is walked rather than dropped, and that is the whole reason
 * this function exists: `fetch` rejects with `TypeError: fetch failed`, and
 * everything that says WHICH failure it was — the refused address, the DNS
 * answer, the timeout — sits one level down in `cause`. Logging only the top
 * message would produce a line that is safe because it says nothing.
 *
 * Bounded at five links. A cycle in `cause` is possible and costs nothing to
 * rule out here.
 */
export function errorText(err: unknown): string {
  const parts: string[] = [];
  let current: unknown = err;

  while (current instanceof Error && parts.length < 5) {
    parts.push(current.message);
    current = current.cause;
  }

  if (parts.length > 0) return parts.join(": ");
  // Not String(err): a thrown object with the default toString renders as
  // "[object Object]", which is noise wearing the shape of information.
  return typeof err === "string" ? err : "non-error thrown";
}

/**
 * Replaces every email and IP address in one left-to-right pass.
 *
 * ONE pass, and the Go fuzzer is why. Doing emails and then addresses as two
 * sweeps was correct on every input anybody thinks of and wrong on two nobody
 * would: `a@b.tld@c.tld` left the second domain standing, and in `0@::0.XA` the
 * IP redaction BUILT an email that was not there. Walking once fixes the class
 * rather than the two cases — what this writes is never read again, so no
 * substitution can create a match.
 */
function redactAddresses(s: string): string {
  // The overwhelmingly common case is a line with none of this in it. One pass
  // to find out costs less than the walk below.
  if (!s.includes("@") && !s.includes(".") && !s.includes(":")) return s;

  let out = "";
  let i = 0;

  while (i < s.length) {
    if (runStart(s, i, isLocalChar)) {
      const end = matchEmail(s, i);
      if (end > 0) {
        out += REDACTED_EMAIL;
        i = end;
        continue;
      }
    }

    // No run-start guard on this one, and the Go fuzzer took it away: it assumed
    // that if the whole run fails to parse, no part of it can. That is false. In
    // "::0X%::0" the second run begins at '%', because '%' is a legal zone
    // separator — "%::0" does not parse, and the "::0" inside it was never
    // tried. An address survived in the output.
    //
    // The cost is a rescan from each position inside a run that failed. The
    // values reaching here are bounded, and correctness is not the thing to
    // trade for it.
    const end = matchAddr(s, i);
    if (end > 0) {
      out += REDACTED_IP;
      i = end;
      continue;
    }

    out += s[i];
    i++;
  }

  return out;
}

/**
 * Whether i begins a run of characters the predicate accepts.
 *
 * Safe for the email matcher and only for it: the local part is scanned greedily
 * and the '@' after it sits at a fixed offset, so starting later can only
 * shorten a match that would be found anyway. It is NOT safe for addresses — see
 * redactAddresses.
 */
function runStart(s: string, i: number, inClass: (c: string) => boolean): boolean {
  return inClass(s[i]) && (i === 0 || !inClass(s[i - 1]));
}

/**
 * How far an email address starting at i reaches, or -1.
 *
 * Deliberately looser than RFC 5321. This is not validating an address somebody
 * typed, it is finding one in a sentence somebody else's server wrote, and the
 * cost of being wrong runs one way: a redacted token that was not an address is
 * a small loss of detail, a missed address is a promise broken.
 */
function matchEmail(s: string, i: number): number {
  let at = i;
  while (at < s.length && isLocalChar(s[at])) at++;

  // No local part, or no '@' after it.
  if (at === i || at >= s.length || s[at] !== "@") return -1;

  let end = at + 1;
  while (end < s.length && isDomainChar(s[end])) end++;

  // An '@' right after the domain means the boundary was not clean —
  // "a@b.tld@c.tld". Swallow it and keep going rather than stop: consuming one
  // token too many costs a word, stopping early leaves half an address.
  while (end < s.length && s[end] === "@") {
    end++;
    while (end < s.length && isDomainChar(s[end])) end++;
  }

  // A trailing dot is punctuation, not domain: "…from a@b.tld." ends a sentence.
  while (end > at + 1 && s[end - 1] === ".") end--;

  if (!isDomainish(s.slice(at + 1, end))) return -1;
  return end;
}

/**
 * How far an IP address starting at i reaches, or -1.
 *
 * The candidate is cut out by shape and handed to a parser rather than matched
 * by a pattern. That is what keeps a timestamp out of it: "11:19:35" looks like
 * an IPv6 address to a regular expression and does not parse as one.
 *
 * Longest first, then shorter — and that second half is not tidiness. The
 * maximal run is often not an address while a prefix of it is: in "::0.::0" the
 * whole run parses as nothing, and trying only the whole run left the leading
 * "::0" in the output one character at a time.
 */
function matchAddr(s: string, i: number): number {
  let end = i;
  while (end < s.length && isAddrChar(s[end])) end++;
  if (end === i) return -1;

  // An IPv4 needs a dot and an IPv6 needs a colon, so a run with neither cannot
  // contain either — and no prefix of it can. This is what keeps the shrinking
  // loop off the values every line carries: a 32-character request id is hex and
  // nothing else, and it leaves here without a single parse.
  const run = s.slice(i, end);
  if (!run.includes(".") && !run.includes(":")) return -1;

  // The shrinking loop below is bounded, and the Go fuzzer is why it had to be.
  // Without this, a run of n characters costs n parses at each of n positions —
  // 2700 characters of colons and digits took SEVEN SECONDS in the Go logger,
  // and a request path is not length-limited. The filter that exists to protect
  // the log would have been the way to stall the service.
  //
  // Sixty-four is comfortably above the longest thing a parser accepts that this
  // service can see: "[ffff:…:255.255.255.255]:65535" is 53 characters.
  if (end > i + MAX_ADDR_LEN) end = i + MAX_ADDR_LEN;

  for (; end > i; end--) {
    if (!worthParsing(s, i, end)) continue;
    if (parsesAsAddr(s.slice(i, end))) return end;
  }

  return -1;
}

/**
 * Whether a candidate ending at `end` can be an address at all, so that the loop
 * in matchAddr does not spend a parse on a length that cannot be one.
 *
 * A trailing dot is punctuation. A trailing colon is punctuation TOO — except
 * when it is the second half of "::", which is IPv6 syntax and part of the
 * address rather than after it.
 *
 * THIS IS A REPAIR, NOT A PORT. api/internal/logx/scrub.go skipped every
 * candidate ending in a colon, so "peer 2001:db8:: is gone" came out of Scrub
 * unchanged — an address in the log of a service whose privacy page promises
 * none. The Go fuzzer could not find it: its property asks whether the filter
 * still sees an address in its OWN output, and a candidate the matcher never
 * looks at is one the property never asks about. scrub.test.ts checks the output
 * with net.isIP over every substring instead, which shares no code with the
 * matcher, and produced the input in three thousand random cases. The Go side
 * carries the same repair.
 */
function worthParsing(s: string, i: number, end: number): boolean {
  const c = s[end - 1];
  if (c === ".") return false;
  if (c === ":") return end - 2 >= i && s[end - 2] === ":";
  return true;
}

/**
 * Whether one candidate is an address, with or without a port.
 */
function parsesAsAddr(token: string): boolean {
  if (addrKind(token) !== 0) return true;

  // "[::1]:8080". Brackets mean IPv6 and mean a port follows.
  if (token.startsWith("[")) {
    const close = token.indexOf("]");
    if (close < 2 || close + 1 >= token.length || token[close + 1] !== ":") return false;
    return isPort(token.slice(close + 2)) && addrKind(token.slice(1, close)) === 6;
  }

  // "127.0.0.1:8080". The port goes with it: an address is not less identifying
  // for having one.
  const colon = token.lastIndexOf(":");
  if (colon < 1) return false;
  return isPort(token.slice(colon + 1)) && addrKind(token.slice(0, colon)) === 4;
}

/**
 * 4, 6 or 0 — the same answer `net.isIP` gives, written out here.
 *
 * WHY NOT node:net, which already does this. Because importing it makes this
 * module Node-only, and Next builds `instrumentation.ts` for the edge runtime
 * too whether or not this application has an edge route — so the import turned
 * every production build into a warning about a module that cannot load there.
 *
 * The rewrite is worth more than it costs, and not because of the build: it
 * turns `net.isIP` into an INDEPENDENT oracle. scrub.test.ts rescans every
 * output with it, so the test now disagrees with this function on any input
 * where the two read the address differently, over four hundred thousand
 * generated cases. While this file called isIP, the test and the code shared
 * their idea of what an address is, and only the scanning strategy was checked.
 *
 * Deliberately identical to isIP where it matters and never looser: a candidate
 * this accepts and isIP does not is over-redaction, which costs a word; the
 * other direction is a promise broken, and the test fails on it.
 *
 * Exported for that test alone. The boundary is the thing worth checking
 * directly, and api/internal/logx exports Scrub for the same reason.
 */
export function addrKind(token: string): 0 | 4 | 6 {
  const percent = token.indexOf("%");
  if (percent >= 0) {
    // A zone, and only IPv6 carries one. It has to be non-empty and may not
    // contain a second '%' or a bracket — measured against isIP rather than
    // guessed, and it matters: without the rule this accepted 13000 non-addresses
    // per three million generated tokens. Over-redaction costs a word rather than
    // a promise, but a filter that redacts noise is one somebody switches off.
    const zone = token.slice(percent + 1);
    if (zone === "" || /[%[\]]/.test(zone)) return 0;
    return isIPv6(token.slice(0, percent)) ? 6 : 0;
  }
  if (token.includes(":")) return isIPv6(token) ? 6 : 0;
  return isIPv4(token) ? 4 : 0;
}

function isIPv4(token: string): boolean {
  const parts = token.split(".");
  if (parts.length !== 4) return false;
  for (const part of parts) {
    if (!isOctet(part)) return false;
  }
  return true;
}

function isOctet(part: string): boolean {
  if (part.length === 0 || part.length > 3) return false;
  for (const c of part) {
    if (c < "0" || c > "9") return false;
  }
  // "01.2.3.4" is not an address anywhere that matters, and accepting it would
  // let one address be spelled several ways.
  if (part.length > 1 && part.startsWith("0")) return false;
  return Number(part) <= 255;
}

function isIPv6(token: string): boolean {
  let head = token;
  let tail = "";
  let compressed = false;

  const zeros = token.indexOf("::");
  if (zeros >= 0) {
    // One zero run only. Two would leave the length ambiguous, which is the
    // whole reason the notation allows exactly one.
    if (token.includes("::", zeros + 1)) return false;
    compressed = true;
    head = token.slice(0, zeros);
    tail = token.slice(zeros + 2);
  }

  const headParts = head === "" ? [] : head.split(":");
  const tailParts = tail === "" ? [] : tail.split(":");
  const parts = [...headParts, ...tailParts];

  // A group count, not a part count: a trailing IPv4 occupies two.
  let groups = parts.length;

  const last = parts[parts.length - 1];
  if (parts.length > 0 && last.includes(".")) {
    // An embedded IPv4 is the last 32 bits of the address, so nothing may follow
    // it — a zero run included. "0.0.0.0::" reads as one until this line.
    if (compressed && tailParts.length === 0) return false;
    if (!isIPv4(last)) return false;
    groups += 1;
    parts.pop();
  }

  for (const part of parts) {
    if (!isHexGroup(part)) return false;
  }

  // With a zero run there has to be at least one group left to compress;
  // without one, every group is spelled out.
  return compressed ? groups <= 7 : groups === 8;
}

function isHexGroup(part: string): boolean {
  if (part.length === 0 || part.length > 4) return false;
  for (const c of part) {
    if (!((c >= "0" && c <= "9") || (c >= "a" && c <= "f") || (c >= "A" && c <= "F"))) {
      return false;
    }
  }
  return true;
}

function isPort(s: string): boolean {
  if (s.length === 0 || s.length > 5) return false;
  for (const c of s) {
    if (c < "0" || c > "9") return false;
  }
  return Number(s) <= 65535;
}

/**
 * Rejects the half of the '@' cases that are not addresses: a domain needs a dot
 * and a last label that reads like a suffix.
 */
function isDomainish(domain: string): boolean {
  const dot = domain.lastIndexOf(".");
  if (dot < 1 || dot === domain.length - 1) return false;

  const tld = domain.slice(dot + 1);
  if (tld.length < 2) return false;

  for (const c of tld) {
    if (!isAlpha(c)) return false;
  }
  return true;
}

function isAlpha(c: string): boolean {
  return (c >= "a" && c <= "z") || (c >= "A" && c <= "Z");
}

function isAlnum(c: string): boolean {
  return isAlpha(c) || (c >= "0" && c <= "9");
}

function isLocalChar(c: string): boolean {
  return isAlnum(c) || ".!#$%&'*+/=?^_`{|}~-".includes(c);
}

function isDomainChar(c: string): boolean {
  return isAlnum(c) || c === "." || c === "-";
}

/**
 * Every character that can appear in an IPv4 or IPv6 address, its zone, or a
 * bracketed host:port.
 */
function isAddrChar(c: string): boolean {
  return (
    (c >= "0" && c <= "9") ||
    (c >= "a" && c <= "f") ||
    (c >= "A" && c <= "F") ||
    c === "." ||
    c === ":" ||
    c === "[" ||
    c === "]" ||
    c === "%"
  );
}

File: api/internal/logx/scrub.go
package logx

import (
	"fmt"
	"log/slog"
	"net/netip"
	"strings"
)

// What a redacted value is replaced with. A marker rather than an empty string:
// "this line had an address and the address was removed" is a different fact
// from "this line had no address", and only the first one tells you the filter
// is working.
// The marker is built only from characters that may appear inside a domain, and
// it always contains a hyphen. That is not cosmetic, and the fuzzer is why it is
// written down: with brackets around it, redacting an address in the middle of a
// longer token ENDED the domain scan at the bracket and left the text before it
// looking like a valid domain — so a second pass redacted more than the first,
// and "what this function removes" depended on how often it had run. A marker
// the scanner reads straight through cannot move a boundary, and the hyphen
// keeps the result from ever being a well-formed suffix.
const (
	redactedEmail = "redacted-email"
	redactedIP    = "redacted-ip"
)

// Scrub removes control characters, email addresses and IP addresses from one
// string.
//
// Exported because the boundary is worth testing directly, and because F11 is
// told by the build plan to scrub frontend telemetry "wie F1" — when that
// arrives it should call this rather than write a second answer.
func Scrub(s string) string {
	return redactAddresses(StripControl(s))
}

// StripControl removes the bytes that let one log line pretend to be two.
//
// The JSON handler already escapes them, and that was measured rather than
// assumed: a newline in r.URL.Path comes out as an escape INSIDE the string and
// the line stays one line. So this is not what makes forging impossible today —
// it is what keeps it impossible when the writer changes. F2 sends these lines
// through Alloy into Loki and F11 adds a second producer, and "we are safe
// because of how the encoder at the bottom happens to behave" is a guarantee
// living in a different file from the values it protects.
//
// It also settles a recurring argument. CodeQL raises go/log-injection on every
// line that logs a value derived from a request — three of them on this
// service's own pre-existing code — and "the encoder escapes it" is an answer
// somebody has to derive again each time. Removing the characters is an answer
// that reads itself.
//
// A space, not a deletion: two tokens run together read as one value that was
// never there.
func StripControl(s string) string {
	if !hasControl(s) {
		return s
	}

	b := []byte(s)
	for i := range b {
		if isControl(b[i]) {
			b[i] = ' '
		}
	}
	return string(b)
}

func hasControl(s string) bool {
	for i := range len(s) {
		if isControl(s[i]) {
			return true
		}
	}
	return false
}

// isControl covers C0 and DEL. Multi-byte UTF-8 is left alone: every
// continuation byte is >= 0x80, so this cannot cut a rune in half.
func isControl(c byte) bool { return c < 0x20 || c == 0x7F }

// redactAddresses replaces every email and IP address in one left-to-right pass.
//
// ONE pass, and the fuzzer is why. Doing emails and then addresses as two
// separate sweeps was correct on every input anybody thought of and wrong on a
// shape nobody would: replacing an address could leave the marker sitting next
// to a domain, and that token then read as an address ITSELF on the next sweep.
// So the answer depended on how many times the function had run, and the first
// run had left part of an address standing. Two examples the fuzzer minimised
// down to, both real if contrived:
//
//	a@b.tld@c.tld   →  the second domain survived the first sweep
//	0@::0.XA        →  redacting the IPv6 built an email that was not there
//
// Walking once fixes the class rather than the two cases: what this writes is
// never read again, so no substitution can create a match.
//
// A match is only attempted at the START of a run. That keeps the scan linear
// and stops a suffix of a longer token from matching on its own.
func redactAddresses(s string) string {
	// The overwhelmingly common case is a line with none of this in it. One
	// pass to find out costs less than the walk below.
	if !strings.ContainsAny(s, "@.:") {
		return s
	}

	var b strings.Builder
	b.Grow(len(s))

	for i := 0; i < len(s); {
		if runStart(s, i, isLocalByte) {
			if end, ok := matchEmail(s, i); ok {
				b.WriteString(redactedEmail)
				i = end
				continue
			}
		}
		// No run-start guard on this one, and the fuzzer took it away: it
		// assumed that if the whole run fails to parse, no part of it can. That
		// is false. In "::0X%::0" the second run begins at '%', because '%' is
		// a legal zone separator — "%::0" does not parse, and the "::0" inside
		// it was never tried. An address survived in the output.
		//
		// The cost is a rescan from each position inside a run that failed, so
		// a run of n bytes is O(n²) in the worst case. The values reaching here
		// are bounded — a relay message is cut to 200 characters before it ever
		// becomes an error — and correctness is not the thing to trade for it.
		if end, ok := matchAddr(s, i); ok {
			b.WriteString(redactedIP)
			i = end
			continue
		}
		b.WriteByte(s[i])
		i++
	}

	return b.String()
}

// runStart reports whether i begins a run of bytes the predicate accepts.
//
// Safe for the email matcher and only for it: the local part is scanned
// greedily and the '@' after it sits at a fixed offset, so starting later can
// only shorten a match that would be found anyway. It is NOT safe for
// addresses — see redactAddresses.
func runStart(s string, i int, in func(byte) bool) bool {
	return in(s[i]) && (i == 0 || !in(s[i-1]))
}

// matchEmail reports how far an address starting at i reaches.
//
// Deliberately looser than RFC 5321. This is not validating an address somebody
// typed, it is finding one in a sentence somebody else's server wrote, and the
// cost of being wrong runs one way: a redacted token that was not an address is
// a small loss of detail, a missed address is a promise broken.
func matchEmail(s string, i int) (int, bool) {
	at := i
	for at < len(s) && isLocalByte(s[at]) {
		at++
	}
	// No local part, or no '@' after it.
	if at == i || at >= len(s) || s[at] != '@' {
		return 0, false
	}

	end := at + 1
	for end < len(s) && isDomainByte(s[end]) {
		end++
	}

	// An '@' right after the domain means the boundary was not clean —
	// "a@b.tld@c.tld". Swallow it and keep going rather than stop: consuming
	// one token too many costs a word, stopping early leaves half an address.
	for end < len(s) && s[end] == '@' {
		end++
		for end < len(s) && isDomainByte(s[end]) {
			end++
		}
	}

	// A trailing dot is punctuation, not domain: "…from a@b.tld." ends a
	// sentence.
	for end > at+1 && s[end-1] == '.' {
		end--
	}

	if !isDomainish(s[at+1 : end]) {
		return 0, false
	}
	return end, true
}

// matchAddr reports how far an IP address starting at i reaches.
//
// The candidate is cut out by shape and handed to net/netip rather than matched
// by a pattern. That is what keeps a timestamp out of it: "11:19:35" looks like
// an IPv6 address to a regular expression and does not parse as one.
//
// Longest first, then shorter — and that second half is not tidiness. The
// maximal run is often not an address while a prefix of it is: in "::0.::0" the
// whole run parses as nothing, and trying only the whole run left the leading
// "::0" in the output one byte at a time. The fuzzer found it; reading did not.
func matchAddr(s string, i int) (int, bool) {
	end := i
	for end < len(s) && isAddrByte(s[end]) {
		end++
	}
	if end == i {
		return 0, false
	}

	// An IPv4 needs a dot and an IPv6 needs a colon, so a run with neither
	// cannot contain either — and no prefix of it can. This is what keeps the
	// shrinking loop below off the values every line carries: a 32-character
	// request id is hex and nothing else, and it leaves here without a single
	// call into net/netip.
	if !strings.ContainsAny(s[i:end], ".:") {
		return 0, false
	}

	// The shrinking loop below is bounded, and the fuzzer is why it had to be.
	//
	// Without this, a run of n bytes costs n parses at each of n positions. A
	// 2700-character run of colons and digits took SEVEN SECONDS in the
	// logger — and r.URL.Path is not length-limited, so a stranger could have
	// spent this service's CPU by asking for a long enough path. The filter
	// that exists to protect the log would have been the way to stall it.
	//
	// Sixty-four is comfortably above the longest thing net/netip accepts that
	// this service can see: "[ffff:…:255.255.255.255]:65535" is 53 characters.
	// What it excludes is an IPv6 zone name longer than the address, and a
	// container talking to another container over TCP does not have one.
	if end > i+maxAddrLen {
		end = i + maxAddrLen
	}

	for ; end > i; end-- {
		if !worthParsing(s, i, end) {
			continue
		}

		token := s[i:end]
		if _, err := netip.ParseAddr(token); err == nil {
			return end, true
		}
		// "127.0.0.1:8080" and "[::1]:8080". The port goes with it: an address
		// is not less identifying for having one.
		if _, err := netip.ParseAddrPort(token); err == nil {
			return end, true
		}
	}

	return 0, false
}

// isDomainish rejects the half of the '@' cases that are not addresses: a
// domain needs a dot and a last label that reads like a suffix.
func isDomainish(domain string) bool {
	dot := strings.LastIndexByte(domain, '.')
	if dot < 1 || dot == len(domain)-1 {
		return false
	}
	tld := domain[dot+1:]
	if len(tld) < 2 {
		return false
	}
	for i := range len(tld) {
		c := tld[i]
		if (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') {
			return false
		}
	}
	return true
}

func isLocalByte(c byte) bool {
	switch {
	case c >= 'a' && c <= 'z', c >= 'A' && c <= 'Z', c >= '0' && c <= '9':
		return true
	}
	return strings.IndexByte(".!#$%&'*+/=?^_`{|}~-", c) >= 0
}

func isDomainByte(c byte) bool {
	switch {
	case c >= 'a' && c <= 'z', c >= 'A' && c <= 'Z', c >= '0' && c <= '9':
		return true
	}
	return c == '.' || c == '-'
}

// worthParsing reports whether a candidate ending at end can be an address at
// all, so that the loop in matchAddr does not spend a parse on a length that
// cannot be one.
//
// A trailing dot is punctuation. A trailing colon is punctuation TOO — except
// when it is the second half of "::", which is IPv6 syntax and part of the
// address rather than after it. Skipping those left every address ending in a
// zero run standing: "peer 2001:db8:: is gone" came out of Scrub unchanged.
//
// The fuzzer could not have found this, and that is the part worth writing
// down. FuzzScrubRemovesEveryAddressItCanSee asks whether the filter still sees
// an address in its own output — so a candidate this matcher never looks at is
// one the property never asks about. The web port in F1b checks the output with
// net.isIP over every substring instead, which shares no code with the matcher,
// and it produced this input in three thousand random cases.
func worthParsing(s string, i, end int) bool {
	switch s[end-1] {
	case '.':
		return false
	case ':':
		return end-2 >= i && s[end-2] == ':'
	}
	return true
}

// maxAddrLen bounds one candidate. See matchAddr for what it costs and why.
const maxAddrLen = 64

// isAddrByte covers every byte that can appear in an IPv4 or IPv6 address, its
// zone, or a bracketed host:port.
func isAddrByte(c byte) bool {
	switch {
	case c >= '0' && c <= '9', c >= 'a' && c <= 'f', c >= 'A' && c <= 'F':
		return true
	}
	return c == '.' || c == ':' || c == '[' || c == ']' || c == '%'
}

// scrubValue walks one attribute value.
//
// Three kinds carry text and are handled; the rest are numbers, times and
// booleans, which cannot hold an address. A struct logged with slog.Any could,
// and is deliberately not walked — nothing in this service logs one, and a
// reflective walk of arbitrary values is a lot of machinery guarding a case
// that does not exist. If one appears, this is where it goes, and the ADR says
// so.
func scrubValue(v slog.Value) slog.Value { return walk(v, Scrub) }

// walk applies one string transform to every value that carries text.
func walk(v slog.Value, f func(string) string) slog.Value {
	switch v.Kind() {
	case slog.KindString:
		if s := f(v.String()); s != v.String() {
			return slog.StringValue(s)
		}
		return v

	case slog.KindGroup:
		attrs := v.Group()
		out := make([]slog.Attr, 0, len(attrs))
		changed := false
		for _, a := range attrs {
			next := a
			next.Value = walk(a.Value.Resolve(), f)
			if !next.Value.Equal(a.Value) {
				changed = true
			}
			out = append(out, next)
		}
		if !changed {
			return v
		}
		return slog.GroupValue(out...)

	case slog.KindAny:
		// The path that matters. internal/mail wraps the relay's own words into
		// an error, internal/contact logs that error, and a 550 from a relay
		// routinely quotes the address it refused. The value renders as its
		// Error() string, so that is what gets filtered.
		switch a := v.Any().(type) {
		case error:
			if s := f(a.Error()); s != a.Error() {
				return slog.StringValue(s)
			}
		case fmt.Stringer:
			if s := f(a.String()); s != a.String() {
				return slog.StringValue(s)
			}
		}
		return v
	}

	return v
}

// selfAuthored names the attribute keys whose values this service generates
// itself and whose shape it therefore knows.
//
// There is exactly one, and it was found by running the thing rather than by
// reasoning about it: an RFC 5322 Message-ID is an addr-spec — `msg_01M0PZ…@timseil.dev`
// — so it IS address-shaped, and the filter replaced the whole of it with the
// marker. The line that logs it exists to prove a message was assembled, and it
// was left saying that an address had been removed instead. A reader would have
// concluded the opposite of the truth.
//
// The entry is a claim about the VALUE, not a convenience for the key: this id
// is built in internal/contact from an identifier this service minted plus its
// own domain, and no part of it comes from a visitor. Adding a key here means
// making that claim again, and it is the kind of claim that gets a line in a
// review.
var selfAuthored = map[string]bool{"message_id": true}

func scrubAttr(a slog.Attr) slog.Attr {
	if selfAuthored[a.Key] {
		// Exempt from REDACTION, not from the filter. The entry above is a
		// claim about the SHAPE of the value and says nothing about control
		// characters — a key nobody may redact is still not a key anyone may
		// use to write a second line.
		a.Value = walk(a.Value.Resolve(), StripControl)
		return a
	}
	a.Value = scrubValue(a.Value.Resolve())
	return a
}
